R/write.R

```diff
@@ -149,56 +215,298 @@ add_labelled_xlsx_features <- function(supertbl,
+#'
+#' @keywords internal
+
+check_labelled <- function(supertbl, labelled, call = caller_env()) {
+  # supertbl is considered labelled if cols have label attributes
+  is_labelled <- some(supertbl, function(x) !is.null(attr(x, "label")))
+
+  # If user declared labelled is FALSE return FALSE
+  if (!is.null(labelled) && !labelled) {
+    return(FALSE)
+  }
+
+  # If user not declared and no labels detected, return FALSE
+  if (is.null(labelled) && !is_labelled) {
+    return(FALSE)
+  }
+
+  # If labels are detected & labelled is installed, return TRUE
+  if (is_labelled && is_installed("labelled")) {
+    return(TRUE)
+  }
+
+  # If labels detected and labelled is not installed, return warning message and FALSE
+  if (is_labelled && !is_installed("labelled")) {
+    cli_warn(
+      message = c(
+        "!" = "Labels detected, but {.pkg labelled} not installed. Labels not applied.",
+        "i" = "Consider installing {.pkg labelled} and re-running."
+      ),
+      call = call
+    )
+    return(FALSE)
+  }
+
+  # Otherwise error, meaning labels asked for on a non-labelled input
+  cli_abort(
+    message = c(
+      "x" = "{.arg labelled} declared TRUE, but no labels detected.",
+      "i" = "Was {.fun REDCapTidieR::make_labelled() used first?}"
+    ),
+    call = call
+  )
+}
+
+
+#' @title Recode fields using supertbl metadata
+#'
+#' @description
+#' This utility function helps to map metadata field types in order to apply
+#' changes in supertbl tables.
+#'
+#' @param supertbl A supertibble generated using `read_redcap()`
+#' @param superbl_meta an `unnest`-ed metadata tibble from the supertibble
+#' @param recode_yn Convert REDCap yesno questions from TRUE/FALSE to "yes"/"no"
+#' for readability. Default TRUE.
+#'
+#' @importFrom dplyr mutate across case_when filter pull
+#' @importFrom purrr map
+#' @importFrom tidyselect any_of
+#'
+#' @keywords internal
+
+supertbl_recode <- function(supertbl, supertbl_meta, recode_yn) {
+
+  # Recode yesno from TRUE/FALSE to "yes"/"no"
+  if (recode_yn) {
+    yesno_fields <- supertbl_meta %>%
+      filter(.data$field_type == "yesno") %>%
+      pull(.data$field_name)
+
+    supertbl$redcap_data <- map(
+      supertbl$redcap_data,
+      function(x) .x <- x %>% #nolint: object_usage_linter
+        mutate(across(any_of(yesno_fields),
+                      ~case_when(. == "TRUE" ~ "yes",
+                                 . == "FALSE" ~ "no",
+                                 TRUE ~ NA_character_)))
+    )
+  }
+
+  supertbl$redcap_data
+}
+
+#' @title Bind supertbl metadata
+#'
+#' @description
+#' Simple helper function for binding supertbl metadata into one table. This
+#' supports creating the metadata XLSX sheet as well as `supertbl_recode`.
+#'
+#' @param supertbl A supertibble generated using `read_redcap()`
+#'
+#' @importFrom dplyr filter select
+#' @importFrom tidyr unnest
+#'
+#' @keywords internal
+
+bind_supertbl_metadata <- function(supertbl) {
+  supertbl %>%
```

Comment:
Just checking - does the resulting metadata sheet show which instrument each variable is coming from?

Reply:
Not currently, no. The form name isn't included since it's in the top level supertbl and then everything get's bound. Definitely doable though.